src/main/kotlin/cacophony/controlflow/generation/CFGGenerator.kt

```diff
@@ -4,47 +4,40 @@ import cacophony.controlflow.*
 import cacophony.controlflow.functions.CallGenerator
 import cacophony.controlflow.functions.FunctionHandler
 import cacophony.semantic.analysis.UseTypeAnalysisResult
+import cacophony.semantic.analysis.VariablesMap
 import cacophony.semantic.names.ResolvedVariables
-import cacophony.semantic.syntaxtree.Block
-import cacophony.semantic.syntaxtree.Definition
-import cacophony.semantic.syntaxtree.Empty
-import cacophony.semantic.syntaxtree.Expression
-import cacophony.semantic.syntaxtree.FunctionCall
-import cacophony.semantic.syntaxtree.Literal
-import cacophony.semantic.syntaxtree.OperatorBinary
-import cacophony.semantic.syntaxtree.OperatorUnary
-import cacophony.semantic.syntaxtree.Statement
-import cacophony.semantic.syntaxtree.VariableUse
+import cacophony.semantic.syntaxtree.*
+import cacophony.semantic.types.TypeCheckingResult
 
 /**
  * Converts Expressions into CFG
  */
 internal class CFGGenerator(
     private val resolvedVariables: ResolvedVariables,
-    analyzedUseTypes: UseTypeAnalysisResult, // TODO: adjust to new specification of analyzedUseTypes
+    analyzedUseTypes: UseTypeAnalysisResult,
     private val function: Definition.FunctionDefinition,
     private val functionHandlers: Map<Definition.FunctionDefinition, FunctionHandler>,
+    val variablesMap: VariablesMap,
+    private val typeCheckingResult: TypeCheckingResult,
     private val callGenerator: CallGenerator,
 ) {
     private val cfg = CFG()
     private val sideEffectAnalyzer = SideEffectAnalyzer(analyzedUseTypes)
     private val operatorHandler = OperatorHandler(cfg, this, sideEffectAnalyzer)
-    private val assignmentHandler = AssignmentHandler(cfg, this)
+    private val assignmentHandler = AssignmentHandler(this)
     private val prologue = listOfNodesToExtracted(getCurrentFunctionHandler().generatePrologue())
     private val epilogue = listOfNodesToExtracted(getCurrentFunctionHandler().generateEpilogue())
 
     internal fun generateFunctionCFG(): CFGFragment {
         val bodyCFG = visit(function.body, EvalMode.Value, Context(null))
-        val returnValueRegister = registerUse(getCurrentFunctionHandler().getResultRegister())
+        val returnValueRegister = getCurrentFunctionHandler().getResultLayout()
+        require(returnValueRegister is SimpleLayout)
+        require(returnValueRegister.access is CFGNode.LValue)
 
         val extended =
             when (bodyCFG) {
                 is SubCFG.Extracted -> extendWithAssignment(bodyCFG, returnValueRegister, EvalMode.Value)
-                is SubCFG.Immediate -> {
-                    val node = CFGNode.Assignment(returnValueRegister, bodyCFG.access)
-                    val vertex = cfg.addUnconditionalVertex(node)
-                    SubCFG.Extracted(vertex, vertex, returnValueRegister)
-                }
+                is SubCFG.Immediate -> assignLayoutWithValue(bodyCFG.access, returnValueRegister, returnValueRegister)
```

Comment:
I know we are requiring this to be SimpleLayout for now, but it would be better to change name from returnValueRegister

Reply:
A lot of stuff here depends on the prologue generation, so there will be more changes.

Reply:
Name changed, and removed `require()`s. Still, there are gonna be more changes probably, after the prologue and epilogue generation.